File: src/main/java/org/moinex/model/investment/CryptoExchange.java
/*
 * Filename: CryptoExchange.java
 * Created on: January 28, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model.investment;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.moinex.util.Constants;

/**
 * Class that represents a crypto exchange
 */
@Entity
@Table(name = "crypto_exchange")
@Getter
@Setter
@NoArgsConstructor
@SuperBuilder
public class CryptoExchange
{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "sold_crypto_id", referencedColumnName = "id", nullable = false)
    private Ticker soldCrypto;

    @ManyToOne
    @JoinColumn(name                 = "received_crypto_id",
                referencedColumnName = "id",
                nullable             = false)
    private Ticker receivedCrypto;

    @Column(name = "sold_quantity", nullable = false)
    private BigDecimal soldQuantity;

    @Column(name = "received_quantity", nullable = false)
    private BigDecimal receivedQuantity;

    @Column(name = "date", nullable = false)
    private String date;

    @Column(name = "description" )
    private String description;

    public abstract static class CryptoExchangeBuilder<
        C extends CryptoExchange, B extends CryptoExchangeBuilder<C, B>>
    {
        public B date(LocalDateTime date)
        {
            this.date = date.format(Constants.DB_DATE_FORMATTER);
            return self();
        }
    }

    /**
     * Constructor for testing purposes
     */
    public CryptoExchange(Long          id,
                          Ticker        soldCrypto,
                          Ticker        receivedCrypto,
                          BigDecimal    soldQuantity,
                          BigDecimal    receivedQuantity,
                          LocalDateTime date,
                          String        description)
    {
        this.id               = id;
        this.soldCrypto       = soldCrypto;
        this.receivedCrypto   = receivedCrypto;
        this.soldQuantity     = soldQuantity;
        this.receivedQuantity = receivedQuantity;
        this.date             = date.format(Constants.DB_DATE_FORMATTER);
        this.description      = description;
    }

    public LocalDateTime getDate()
    {
        return LocalDateTime.parse(date, Constants.DB_DATE_FORMATTER);
    }

    public void setDate(LocalDateTime date)
    {
        this.date = date.format(Constants.DB_DATE_FORMATTER);
    }
}

File: src/main/kotlin/org/moinex/model/investment/CryptoExchange.kt
/*
 * Filename: CryptoExchange.kt (original filename: CryptoExchange.java)
 * Created on: January 28, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model.investment

import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.moinex.common.converter.LocalDateTimeStringConverter
import org.moinex.common.extension.toRounded
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "crypto_exchange")
class CryptoExchange(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @ManyToOne
    @JoinColumn(name = "sold_crypto_id", referencedColumnName = "id", nullable = false)
    var soldCrypto: Ticker,
    @ManyToOne
    @JoinColumn(name = "received_crypto_id", referencedColumnName = "id", nullable = false)
    var receivedCrypto: Ticker,
    @Column(name = "sold_quantity", nullable = false)
    var soldQuantity: BigDecimal,
    @Column(name = "received_quantity", nullable = false)
    var receivedQuantity: BigDecimal,
    @Column(name = "date", nullable = false)
    @Convert(converter = LocalDateTimeStringConverter::class)
    var date: LocalDateTime,
    @Column(name = "description")
    var description: String? = null,
) {
    init {
        soldQuantity = soldQuantity.toRounded()
        receivedQuantity = receivedQuantity.toRounded()

        require(soldQuantity > BigDecimal.ZERO) {
            "Sold quantity must be positive"
        }
        require(receivedQuantity > BigDecimal.ZERO) {
            "Received quantity must be positive"
        }
    }

    override fun toString(): String = "CryptoExchange [id=$id, ${soldCrypto.symbol} -> ${receivedCrypto.symbol}]"
}
